### ISSUE TITLE
Game lags when adding new bodies to the screen (only in release mode on real device iPhone6 and iPhone5S)

### ISSUE THREAD

**Issue by ahmdsdk** [bberak/nature-of-code-app#24]
Hi,
Again I want to say thanks for this awesome library. I have been trying to create a simple game where new bodies are added to the screen and there's a main body that tries to catch or avoid incoming bodies. I used React Native Donkey Kong as reference while creating it. Everything works fine on the simulator (even in release mode on simulator), but when I run release mode on real device the game lags whenever a new body is added.

This is my code, please let me know if I am doing anything wrong.

`const { width, height } = Dimensions.get("screen");
const logoSize = Math.trunc(Math.max(width, height) * 0.075);
const engine = Matter.Engine.create({ enableSleeping: false });

engine.world.gravity.y = 0;

const world = engine.world;
const logoBody = Matter.Bodies.rectangle(width / 2, height / 1.15, logoSize, logoSize,
  {
    isStatic: true,
    label: "logo",
    collisionFilter: {
      category: 1,
      mask: 2,
    },
  }
);

Matter.World.add(world, [logoBody]);

const Physics = (entities, { time }) => {
  let engine = entities.physics.engine;
  Matter.Engine.update(engine, time.delta);
  return entities;
};

let boxIds = 0;

let counter = 0;

let speed = 4;

let speedCounter = 0;

let running = true;

let gameOver = false;

let gravity = 1;

let Entities = {
  physics: { engine: engine, world: world },
  logoBody: { body: logoBody, size: [logoSize, logoSize], color: '#daa520', renderer: Box, render: { visible: true, opacity: 1, }, },
  camera: { offsetY: 0 }
};

const MoveLogo = (entities, { touches, screen }) => {

  speed = gameOver ? 4 : speedCounter%2===0 ? speed + 0.01 : speed;

  running && Matter.Body.translate(logoBody, {x: 0, y: -parseFloat(speed).toFixed(2)});

  running && touches.filter(t => t.type === "press").forEach(t => {

    speedCounter += 1;

    if (t.event.pageX > logoBody.bounds.max.x - (logoSize/2) && logoBody.bounds.max.x < width-(width/10)) {

      return Matter.Body.translate(logoBody, {x: Math.trunc(width/3), y: 0});

    } else if (t.event.pageX < logoBody.bounds.min.x + (logoSize/2) && logoBody.bounds.min.x > width/10) {

      return Matter.Body.translate(logoBody, {x: -Math.trunc(width/3), y: 0});

    }
  });
  return entities;
};

const Camera = (entities, { screen, animations }) => {

	let logoBody = entities.logoBody;
	let camera = entities.camera;
	let targetY = logoBody.body.position.y + camera.offsetY;
  let anchorY = screen.height * 0.85;
	let diff = anchorY - logoBody.body.position.y - camera.offsetY;

	camera.offsetY += diff * 0.85;

	return entities;
};

const RemoveBodies = (entities, {events}) => {
  let world = entities.physics.world;

  let camera = entities.camera;

  let logoBody = entities.logoBody;

  let stopped = events.find(e => e.type === "stopped");

  if (stopped) {
    Object.values(entities).forEach(e => {
      // e.body && console.log("STOPPPED", e.body.label);
      e.body && e.body.label !== "logo" && Matter.World.remove(world, e.body);
      // e.body && e.body.label !== "logo" && Matter.Body.scale(e.body, 0.1, 0.1, {x: e.body.bounds.min.x-1000, y: e.body.bounds.max.y});
    });
    logoBody.render.visible = true;
    logoBody.render.opacity = 1;
    Matter.Engine.clear(engine);
    // camera.offsetY = 0;
    gameOver = true;
    running = false;
    entities = Entities;
    boxIds = 0;
    speedCounter = 0;
  }

  let restart = events.find(e => e.type === "restart");

  if (restart) {
    Object.values(entities).forEach(e => {
      // e.body && console.log("RESTARTING", e.body.label);
      e.body && e.body.label !== "logo" && Matter.World.remove(world, e.body);
      e.body && e.body.label !== "logo" && Matter.Body.scale(e.body, 0.1, 0.1, {x: e.body.bounds.min.x-1000, y: e.body.bounds.max.y});
    });
    // camera.offsetY = 0;
    Matter.Engine.clear(engine);
    gameOver = false;
    entities = Entities;
    boxIds = 0;
    speedCounter = 0;
  }

  let cancel = events.find(e => e.type === "cancel");

  if (cancel) {
    Object.values(entities).forEach(e => {
      // e.body && console.log("Cancel", e.body.label);
      e.body && e.body.label !== "logo" && Matter.World.remove(world, e.body);
      e.body && e.body.label !== "logo" && Matter.Body.scale(e.body, 0.1, 0.1, {x: e.body.bounds.min.x-1000, y: e.body.bounds.max.y});
    });
    // camera.offsetY = 0;
    Matter.Engine.clear(engine);
    entities = Entities;
    boxIds = 0;
    speedCounter = 0;
  }

  // console.log("entities", entities);

  return entities;
};

let lastWp = null, lastAp = null, lastBlock = null, lastNull = null; lastGhost = null, lastSunglasses = null, lastLaugh = null;

const Dispatcher = (entities, {events, time, dispatch}) => {

  if (running && lastBlock === null) {
    lastBlock = time.current;
    setTimeout(() => {
      dispatch({type: "block"});
    }, 300);
  }

  if (time.current - lastBlock > 1500) {
    lastBlock = running ? time.current : null;
    running && dispatch({type: "block"});
  }

  if (running && lastAp === null) {
		lastAp = time.current;
		setTimeout(() => {
			dispatch({type: "coin"});
		}, 2000);
	}

  if (time.current - lastAp > 6000) {
    lastAp = running ? time.current : null;
    dispatch({type: "coin"});
  }

  return entities;
};

const AddBodies = (entities, {events, screen}) => {

  let world = entities.physics.world;

  let camera = entities.camera;

  let logoBody = entities.logoBody;

  let coin = events.find(e => e.type === "coin");

	if (coin) {
    let rand = Math.trunc(Math.random()*10);

    let coinBody = Matter.Bodies.circle(
      rand % 3 === 0 ? width/6 : rand % 7 === 1 ? width/2 : width-(width/6),
      -camera.offsetY,
      // 0,
      logoSize/1.25,
      {
        label: rand % 2 === 0 ? "AP-1" : "Spinner",
        collisionFilter: {
          category: 2,
          mask: 1,
        },
      }
    );

    Matter.World.add(world, [coinBody]);

    entities[++boxIds] = {
      body: coinBody,
      size: [logoSize/1.25, logoSize/1.25],
      color: "#daa520",
      renderer: Box,
    };

    counter += 1;
	}

  let block = events.find(e => e.type === "block");

  if (block) {
    let rand = Math.trunc(Math.random()*10);

    let blockBody = Matter.Bodies.rectangle(
      rand % 2 === 0 ? logoBody.body.position.x : rand % 3 === 0 ? width/6 : rand % 5 === 1 ? width/2 : width-(width/6),
      -camera.offsetY,
      // 0,
      logoSize*2,
      logoSize*2/5,
      {
        label: "Block-red",
        collisionFilter: {
          category: 2,
          mask: 1,
        },
      }
    );

    Matter.World.add(world, [blockBody]);

    entities[++boxIds] = {
      body: blockBody,
      size: [logoSize*2, logoSize*2/5],
      color: "red",
      renderer: Box,
    };

    counter += 1;
	}

  return entities;
};

const CameraRenderer = (state, screen) => {
	return (
		<View style={{marginTop: state.camera.offsetY}}>
			{Object.keys(state)
				.filter(key => state[key].renderer)
				.map(key => {
					let entity = state[key];
					if (typeof entity.renderer === "object")
						return (
							<entity.renderer.type
								key={key}
								{...entity}
								screen={screen}
							/>
						);
					else if (typeof entity.renderer === "function")
						return (
							<entity.renderer
								key={key}
								{...entity}
								screen={screen}
							/>
						);
				})}
		</View>
	);
};`

This is the GameEngine component:

`<GameEngine
        ref={g => this.game = g}
        style={styles.container}
        systems={[Physics, MoveLogo, AddBodies, Camera, RemoveBodies, Dispatcher]} // Array of Systems
        entities={Entities}
        running={this.state.running}
        touchProcessor={DefaultTouchProcessor({
          triggerPressEventBefore: 150,
          triggerLongPressEventAfter: 151
        })}
        renderer={CameraRenderer}
        onEvent={this._handleEvent}>
        <StatusBar hidden={true} />
</GameEngine>`

This lag only happens on release mode on a real device iPhone6 and iPhone5S. Release mode on simulator works find. Any help would be highly appreciated. Thank you in advance.

UPDATE: the game works fine on iPhone7, iPhoneX and iPhoneXS Max

**Comment by bberak** [comment]
Hi @ahmdsdk 

Apologies for not replying sooner to your issues - I'm getting a new pc up and running..

Can you share a simplified version of your repo with me so I can run the app on my iPhone SE?

I also have access to a 6S - and can confirm that JS-based gaming performance in general is terrible.

Also, how many bodies are you adding into the world? Have you tried running the RNGE Handbook on your physical device to compare the performance (there is a physics-based example there - similar to what you're trying to achieve).

Does the lag occur only when you add a new body? Or during the whole game? If it only occurs when you add a new body, is there a way for you to recycle bodies (ie, don't remove them entirely, but keep them in some sort of pool for later reuse?).

**Comment by ahmdsdk** [comment]
Hello @bberak 

Thank you for replying. I will try to create a simplified version of my repo and share it with you. I have tried to clone and run both RNKK and the RNGE Handbook but I kept getting some errors on iOS so I just went with reading the code to understand how things work.

I also used this [tutorial](https://medium.com/@williamyang93/my-journey-with-react-native-game-engine-part-i-starting-the-project-bbebcd2ccf6) to figure out how to start with RNGE.

I'm creating and adding bodies on the spot with a timer like you did with the barrels and kong; and the thing is that it starts to lag from the first or second body and whenever a new body is thrown into the world a glitch happens.

I like your idea about recycling bodies, maybe I can try to have a set of bodies already existing from the beginning and then keep recycling them; perhaps, that way I won't need to add any new bodies to the scene and it could solve this lag. I will let you know what happens.

Were you able to take a look at my code above to see if there's anything wrong going on?

Again thanks a lot for your reply =).

**Comment by ahmdsdk** [comment]
Hi again @bberak 

I just tried using only 3 different bodies and recycling them (changing their positions when they fall out of the screen), without adding new bodies to the world and the lag still happened. What do you think the problem could be? Is it just the idea of JS-based gaming not working well with older phones?

**Comment by ahmdsdk** [comment]
@bberak and why do you think the lag happens only on release mode in debugging mode it works fine?

**Comment by bberak** [comment]
Hi @ahmdsdk,

I couldn't find anything blatantly wrong with your code - but I'd still like to try it out in a repo if possible.

There are Expo versions of both RNDK and the RNGE Handbook available - they don't require any compilation or anything, just the Expo app on your phone - you could try these to compare the performance (although there is a bug in RNDK where the image sprites don't appear properly - I haven't fixed that yet).

JS based gaming definitely won't give you the performance of Unreal Engine or Unity - but it is definitely getting better, and still suitable for certain types of games.

Just double checking something - are you saying that you don't have any performance issues when you run your game in Dev mode on a physical device? That would definitely be strange - Release mode should definitely be better performance wise.

The simulator is not always the best way to gauge performance. In my experience it is very temperamental - sometimes it outperforms a physical device - and for certain games it underperforms when compared to a physical device.

**Comment by ahmdsdk** [comment]
Hello @bberak 

Thank you for replying. While creating a simplified  [expo snack](https://snack.expo.io/@ahmad/rnge---test) for you to check out, I kind of was doing some debugging along the way and I realized that the main issue is actually bodies not being removed properly.

What's happening is that they either disappear but still collide with the main body or the game starts to lag. I'm not sure what's the proper way or how to remove those bodies as you can see in that  I kind of tried everything (removing bodies one by one - scaling them to 0.1 - swapping entities). Nothing seems to work.

Now I have the game working properly with no lag even on release mode, but bodies are not being removed as intended and every time I restart the game it is crowded with new and old bodies. Any help regarding this issue is highly appreciated. Thank you again for your time.

P.S. I also encountered an issue with touches of type === "press" (sometimes not working in debug mode but only in release mode - but other times working on both 😅).

**Comment by ahmdsdk** [comment]
Btw here's the [repo](https://github.com/ahmdsdk/rnge--test) of the [snack](https://snack.expo.io/@ahmad/rnge---test).

**Comment by ahmdsdk** [comment]
Hello @bberak 

I just figured out my mistakes. Sorry for wasting your time. I realized that I wasn't properly deleting the bodies after they have fallen out of the screen. I thought that `Matter.World.remove(world, body);` was enough to remove them from the world; although I realized it didn't I kept trying different ways, until I went through your RNDK repo and realized that you were actually deleting the item from the object entities. When I did that all the lagging stopped :D!

Excuse my ignorance but I have just started recently with Matter js and RNGE. I also decided not to go with the `Matter.Events.on()` subscription but instead followed your method using `Matter.Query.ray()`. Thanks again for your time.

On iOS release is working perfect now. I think both these issues have been solved and I can close them now. I'm going to check out the performance on android.

In the future I will fix the repo and the snack to represent the current fixes.

Many thanks again for this awesome library.
